### DISCUSSION TITLE
pnpm v8

### DISCUSSION THREAD

**Discussion by zkochan** [pnpm/pnpm#4967] [General]
This is a discussion about pnpm v8 breaking changes.

- [x] make `auto-install-peers=true` the default. Looks like this is what most people expect since npm has made this the default behavior.
- [x] make `save-workspace-protocol=rolling` the default.
- [x] make `publishConfig.linkDirectory` true by default
- [x] ~make `resolution-mode=time-based` the default~ (too early). But we can start installing the lowest version of direct dependencies.
- [x] - `resolve-peers-from-workspace-root` is `true` by default
- [x] remove the registry fields from the lockfile
- [x] Improve how directory names are shortened. Change `.substring(0, 50)` in [this line](https://github.com/pnpm/pnpm/blob/044cfe517534761d2c1b82f15f0fb9db235be588/packages/dependency-path/src/index.ts#L135) to `.substring(0, 120 - 26 - 1)` 
- [x] set `dedupeDirectDeps` to `true` by default in `@pnpm/core`.
- [x] `resolution-mode` is `lowest-direct` by default.
- [x] `dedupe-peer-dependents` should be `true` by default.
- [x] Drop Node.js 14 support
- [x] Drop support of lockfile v5. Lockfile v5 should be converted to lockfile v6 automatically on `pnpm install`

**Comment by mcmxcdev** [comment]
Not pnpm related per se, but dropping Node v14 + v17 from GitHub Actions and adopting Node 18 (which soon is LTS) should be part of pnpm v8 release.

**Reply by mcmxcdev** [reply]
Related to pnpm config, we tend to set `--ignore-workspace-root-check`to false and `--engine-strict` to true for each project in `.npmrc`. Might be worth discussing.

**Reply by zkochan** [reply]
Regarding Node.js v14 support, it depends on when pnpm v8 will be shipped.

Accidentally installing a package in the root is a frequent mistake, so I am not in favor of allowing it by default.

**Comment by hiepxanh** [comment]
add `pn` as alias for pnpm for faster typing <3

**Comment by EscapeB** [comment]
Change the behavior of `shared-workspace-lockfile=false`

Not like the old behavior by setting `shared-workspace-lockfile=false`, which will have multiple virtual store.

The new expected behavior is we still using the shared virtual store but have individual lock file for each package in pnpm worksapces.

**Comment by mark-omarov** [comment]


**Reply by zkochan** [reply]
It is probably possible but I don't see how it is related to v8 release. Just create a new issue.

**Comment by zkochan** [comment]
I am locking this for contributors only because for some reason people started to make feature requests here.

**Comment by mcmxcdev** [comment]
We use `use-inline-specifiers-lockfile-format=true` since its release in https://github.com/pnpm/pnpm/releases/tag/v7.7.0 and really like it. Having potentially less merge conflicts is a good thing and this could be a contender to be enabled by default in v8.

**Reply by zkochan** [reply]
I think there are some more changes to the lockfile format that we would want to ship.

**Comment by mcmxcdev** [comment]
@zkochan I just noticed that there are still older issues flagged with breaking change around, might make sense to work on them for v8:
https://github.com/pnpm/pnpm/issues?q=is%3Aopen+is%3Aissue+label%3A%22type%3A+breaking+change%22+sort%3Aupdated-desc

**Reply by zkochan** [reply]
This one would be good to do https://github.com/pnpm/pnpm/issues/3054

about the other one I don't care.

**Comment by iobuhov** [comment]
Is there a plan to create migration guide on how to migrate project from 7 to 8?

**Reply by zkochan** [reply]
I am not sure what migration guide would be needed. I think the release page is clear enough: https://github.com/pnpm/pnpm/releases/tag/v8.0.0-rc.0

**Reply by iobuhov** [reply]
Okay

**Comment by dipunm** [comment]
Hey all, I am trying to get my head around one specific feature: `resolve-peers-from-workspace-root`.

From the documentation, I understood it to mean that if any peer dependencies are listed for any dependency of a package, and if the workspace root has that peer dependency installed, then it should indirectly install that peer dependency for the package.

Looking at the `node_modules` this does not seem to happen.

I thought I may have misinterpreted it to mean that if you install the peer dependency, then pnpm will look to the workspace root in order to resolve the correct version number. This is indeed the case, but it seems to work with or without the setting, and with or without `typescript` being listed as a peer dependency.

Another thing I tested was peer dependency version conflicts. If I declare `typescript@~5.1.0` as my peer dependency and install `typescript@5.2.2` in the workspace root, I expect to see some version mismatching warnings.

At this point, I don't think I understand what this feature provides. Can someone help me to understand?

Here is my project layout btw for context:

```
- [root]
  - devDependencies:
    - typescript
  - packages: 
    - ts-config
      - peerDependencies:
        - typescript
    - projectA
      - devDependencies:
        - ts-config
```
